lite/verifying/provider.go

```diff
@@ -0,0 +1,440 @@
+				options.TrustHeight, options.TrustHash, trustCommit.Hash())
+		}
+		return trustCommit, nil
+	} else {
+
+		latestCommit := latestBlock.SignedHeader
+
+		// NOTE: This should really belong in the callback.
+		// WARN THE USER IN ALL CAPS THAT THE LITE CLIENT IS NEW,
+		// AND THAT WE WILL SYNC TO AND VERIFY LATEST COMMIT.
+		fmt.Printf("trusting source at height %v and hash %X...\n", latestCommit.Height, latestCommit.Hash())
+		if options.Callback != nil {
+			err := options.Callback(latestCommit.Height, latestCommit.Hash())
+			if err != nil {
+				return types.SignedHeader{}, err
+			}
+		}
+		return latestCommit, nil
+	}
+}
+
+//----------------------------------------
+
+type nowFn func() time.Time
+
+const sizeOfPendingMap = 1024
+
+var _ lite.UpdatingProvider = (*Provider)(nil)
+
+// Provider implements a persistent caching Provider that
+// auto-validates.  It uses a "source" Provider to obtain the needed
+// FullCommits to securely sync with validator set changes.  It stores properly
+// validated data on the "trusted" local system.
+// NOTE: This Provider can only work with one chainID, provided upon
+// instantiation.
+type Provider struct {
+	chainID     string
+	logger      log.Logger
+	trustPeriod time.Duration // e.g. the unbonding period, or something smaller.
+	now         nowFn
+	height      int64
+
+	// Already validated, stored locally
+	trusted lite.PersistentProvider
+
+	// New info, like a node rpc, or other import method.
+	source lite.Provider
+
+	// pending map to synchronize concurrent verification requests
+	mtx                  sync.Mutex
+	pendingVerifications map[int64]chan struct{}
+}
+
+// makeProvider returns a new verifying Provider. It uses the
+// trusted Provider to store validated data and the source Provider to
+// obtain missing data (e.g. FullCommits).
+//
+// The trusted Provider should be a DBProvider.
+// The source Provider should be a client.HTTPProvider.
+// NOTE: The external facing constructor is called NewVerifyingProivider.
+func makeProvider(chainID string, trustPeriod time.Duration, trusted lite.PersistentProvider, source lite.Provider, logger log.Logger) *Provider {
+	if trustPeriod == 0 {
+		panic("VerifyingProvider must have non-zero trust period")
+	}
+	logger = logger.With("module", loggerPath)
+	trusted.SetLogger(logger)
+	source.SetLogger(logger)
+	return &Provider{
+		logger:               logger,
+		chainID:              chainID,
+		trustPeriod:          trustPeriod,
+		trusted:              trusted,
+		source:               source,
+		pendingVerifications: make(map[int64]chan struct{}, sizeOfPendingMap),
+	}
+}
+
+// Implements Verifier.
+func (vp *Provider) Verify(signedHeader types.SignedHeader) error {
+
+	// We can't verify commits for a different chain.
+	if signedHeader.ChainID != vp.chainID {
+		return cmn.NewError("BaseVerifier chainID is %v, cannot verify chainID %v",
+			vp.chainID, signedHeader.ChainID)
```

Comment:
Nit: We try to avoid using the common package for errors in the tendermint code base.